architecture/core.rst

```diff
@@ -390,6 +390,149 @@ normal world supports scheduling of processes. Even on UP systems, supporting
 several trusted threads in optee_os helps normal world scheduler to be
 efficient.
 
+Core handlers for native interrupts
+===================================
+
+OP-TEE core provides methods for device drivers to setup and register
+handler functions for native interrupt controller drivers
+(see:ref:`native_foreign_irqs`).
+Interrupt handlers can be nested for when an interrupt controller
+exposes interrupts which signaling is multiplexed on an interrupt
+controlled by a parent interrupt controller.
+
+Interrupt controllers are represented by an instance of ``struct itr_chip``.
+An interrupt controller exposes a given number of interrupts, each identified
+by an index from 0 to N-1 where N is the total number of interrupts exposed
+by that controller. In the literature, an interrupt index identifier
+is called interrupt number.
+
+**Interrupt management API functions**
+
+Interrupt management resources are declared in header file interrupt.h_.
+Interrupt consumers main API functions are:
+
+    - ``interrupt_enable()`` and ``interrupt_disable()`` to respectively
+      enable or disable an interrupt.
+
+    - ``interrupt_mask()`` and ``interrupt_unmask()`` to respectively mask
+or unmask an interrupt. Masking of an enabled interrupt temporarily disables
+the interrupt while unmasking enables a previously masked interrupt.
+``interrupt_mask()`` and ``interrupt_unmask()`` are allowed to be called
+from an interrupt context, but ``interrupt_enable()`` and
+``interrupt_disable()`` not so.
+
+    - ``interrupt_configure()`` to configure an interrupt detection mode
+      and priority.
+
+    - ``interrupt_add_handler()`` to configure and interrupt and register
+      a interrupt handler function, see below.
+
+    - ``interrupt_remove_handler()`` to unregister an interrupt handler
+      function from an interrupt.
+
+**Interrupt controller drivers**
+
+An interrupt controller instance, named chip (``struct itr_chip``) defines
+operation function handlers for management of the interrupt(s) it controls.
+An interrupt chip driver must provide operation handler functions ``.add``,
+``.mask``, ``.unmask``, ``.enable`` and ``.disable``. There are other
+operation handler functions that are optional, as ``.rasie_pi``,
+``.raise_sgi`` and ``.set_priority``.
+
+An interrupt chip driver registers the controller instance with API
+function ``itr_chip_init()``. The driver calls the registered interrupt
+consumer(s) handler(s) with API function ``interrupt_call_handlers()``.
+
+**CPU main interrupt controller driver**
+
+CPU main interrupt controller (e.g. a GIC instance on Arm architecture CPUs)
+driver must register itself as main controller using API function
+``interrupt_main_init()``. The function is in charge of calling
+``itr_chip_init()`` for that chip instance.
+
+Interrupt consumer drivers get can get the CPU main interrupt controller
+reference with API function ``interrupt_get_main_chip()``.
```

Comment:
One error (one 'get' too much).
s/... drivers get can get the ... / ... drivers can get the ... /

However, just a suggestion for a slight re-write, leave it up to you to decide.
s/
Interrupt consumer drivers get can get the CPU main interrupt controller reference with API function ...
/
Interrupt consumer drivers can get a reference to the CPU main interrupt controller with the API function ...
/